test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -1620,6 +1693,429 @@ TEST_F(ClusterManagerImplTest, OriginalDstInitialization) {
+
+  // postThreadLocalClusterUpdate() calls.
+  EXPECT_EQ(2, calls);
+}
+
+TEST_F(ClusterManagerImplTest, MergedUpdatesHostChangesInPlace) {
+  createWithLocalClusterUpdate();
+
+  // Ensure we see the right set of added/removed hosts on every call.
+  uint32_t calls = 0;
+  EXPECT_CALL(local_cluster_update_, post(_, _, _))
+      .WillRepeatedly(Invoke([&calls](uint32_t priority, const HostVector& hosts_added,
+                                      const HostVector& hosts_removed) -> void {
+        switch (calls) {
+        case 0:
+          // First immediate call, nothing was added/removed.
+          EXPECT_EQ(0, priority);
+          EXPECT_EQ(0, hosts_added.size());
+          EXPECT_EQ(0, hosts_removed.size());
+          break;
+        case 1:
+          // Merged call, failed/active/failed is merged as 1 call.
+          // Note, note, note: the host gets updated in place in the original
+          // list of hosts in this case, so nothing expected here.
+          EXPECT_EQ(0, priority);
+          EXPECT_EQ(0, hosts_added.size());
+          EXPECT_EQ(0, hosts_removed.size());
+          break;
+        }
+
+        ++calls;
+      }));
+
+  const Cluster& cluster = cluster_manager_->clusters().begin()->second;
+  HostVectorSharedPtr hosts(
+      new HostVector(cluster.prioritySet().hostSetsPerPriority()[0]->hosts()));
+  HostsPerLocalitySharedPtr hosts_per_locality = std::make_shared<HostsPerLocalityImpl>();
+  HostVector hosts_added{};
+  HostVector hosts_removed{};
+
+  // No timer needs to be mocked at this point, since the first update should be applied
+  // immediately.
+  // No added/removed hosts (emulate metadata or health change).
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+
+  // Now we need the timer to be ready, since createTimer() will be call given that this update
+  // _will_ be delayed.
+  Event::MockTimer* timer = new NiceMock<Event::MockTimer>(&factory_.dispatcher_);
+
+  // Host becomes failed (no need to actually change it here).
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+  // Host becomes active.
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+  // Host becomes failed again.
+  cluster.prioritySet().hostSetsPerPriority()[0]->updateHosts(
+      hosts, hosts, hosts_per_locality, hosts_per_locality, {}, hosts_added, hosts_removed);
+
+  // Ensure the merged updates were applied.
+  timer->callback_();
+  EXPECT_EQ(1, factory_.stats_.counter("cluster_manager.merged_updates").value());
+
+  // postThreadLocalClusterUpdate() calls.
+  EXPECT_EQ(2, calls);
+}
+
+TEST_F(ClusterManagerImplTest, MergedUpdatesMetadataLatest) {
```

Comment:
Can you add a one line comment above each of these tests to explain what they do?